phoenix-core/src/main/java/org/apache/phoenix/util/UpgradeUtil.java

```diff
@@ -1279,4 +1304,129 @@ public static boolean truncateStats(HTableInterface metaTable, HTableInterface s
         }
         return false;
     }
+
+    public static void mapTableToNamespace(HBaseAdmin admin, HTableInterface metatable, String srcTableName,
+            String destTableName, ReadOnlyProps props, Long ts, String phoenixTableName, PTableType pTableType)
+                    throws SnapshotCreationException, IllegalArgumentException, IOException, InterruptedException,
+                    SQLException {
+        srcTableName = SchemaUtil.normalizeIdentifier(srcTableName);
+        if (!SchemaUtil.isNamespaceMappingEnabled(
+                SchemaUtil.isSystemTable(srcTableName.getBytes()) ? PTableType.SYSTEM : null,
+                props)) { throw new IllegalArgumentException(SchemaUtil.isSystemTable(srcTableName.getBytes())
+                        ? "For system table " + QueryServices.IS_SYSTEM_TABLE_MAPPED_TO_NAMESPACE
+                                + " also needs to be enabled along with " + QueryServices.IS_NAMESPACE_MAPPING_ENABLED
+                        : QueryServices.IS_NAMESPACE_MAPPING_ENABLED + " is not enabled"); }
+
+        if (PTableType.TABLE.equals(pTableType) || PTableType.INDEX.equals(pTableType)) {
+            admin.snapshot(srcTableName, srcTableName);
+            admin.cloneSnapshot(srcTableName.getBytes(), destTableName.getBytes());
+            admin.disableTable(srcTableName);
```

Comment:
Sounds like we would leaving snapshots dangling around. We should remove snapshots once we have successfully upgraded the table. If the user wants to be cautious he/she can always snapshot their table/s first before running this upgrade. In case of failure, we should try restoring the table/s using the snapshot that Phoenix has taken and attempt to delete  the snapshot after the attempt. We should try and not leavE any residual state from the upgrade, IMHO.